### ISSUE TITLE
Defining a record with empty records deletes the record from PowerDNS remote but not from state file 

### ISSUE THREAD

**Issue by splashx** [terraform-providers/terraform-provider-powerdns#33]
This issue was reported [here](https://github.com/terraform-providers/terraform-provider-powerdns/issues/31#issuecomment-528086027).

# TL;DR:

* Version `1.1.0`:

| | `[]` | `[""]` | 
|---|---|---|
| Record removed from PowerDNS server | ✅  | ✅   |
| Resource removed from state file? | :x:  | ✅  | 
| Apply finished with success? | ✅ | :x: |

* [Proposed patch](https://github.com/terraform-providers/terraform-provider-powerdns/issues/31#issuecomment-528086027)

| | `[]` | `[""]` | 
|---|---|---|
| Record removed from PowerDNS server | ✅  | ✅   |
| Resource removed from state file? | ✅   | ✅  | 
| Apply finished with success? | :x: | :x: |

⚠️ Because we don't implement `Update`, it's important to note that in _all_ cases the record is recreated. So in cases where the record doesn't exist in the state file it's because it failed to create a new one (apply didn't finish with success).

# Long Version

Steps to reproduce:

## Environment with existing record

Environment:
```
$ cat terraform.tfstate
{
  "version": 4,
  "terraform_version": "0.12.7",
  "serial": 24,
  "lineage": "5d15eb7d-2f44-4172-ac9d-494cb2534cd2",
  "outputs": {},
  "resources": []
}
$ cat test.tf
resource "powerdns_record" "foobar" {
  zone    = "sysa.xyz."
  name    = "something.sysa.xyz."
  type    = "A"
  ttl     = 300
  records = ["127.0.0.1"]
}
$ terraform apply  -auto-approve
powerdns_record.foobar: Creating...
powerdns_record.foobar: Creation complete after 0s [id=something.sysa.xyz.:::A]

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.
$ curl -X GET -H "X-Api-Key: secret" http://localhost:8081/api/v1/servers/localhost/zones/sysa.xyz. -s  | jq '.rrsets[].name | match("something.sysa.xyz.")'
{
  "offset": 0,
  "length": 19,
  "string": "something.sysa.xyz.",
  "captures": []
}
$ cat terraform.tfstate
{
  "version": 4,
  "terraform_version": "0.12.7",
  "serial": 26,
  "lineage": "5d15eb7d-2f44-4172-ac9d-494cb2534cd2",
  "outputs": {},
  "resources": [
    {
      "mode": "managed",
      "type": "powerdns_record",
      "name": "foobar",
      "provider": "provider.powerdns",
      "instances": [
        {
          "schema_version": 0,
          "attributes": {
            "id": "something.sysa.xyz.:::A",
            "name": "something.sysa.xyz.",
            "records": [
              "127.0.0.1"
            ],
            "set_ptr": null,
            "ttl": 300,
            "type": "A",
            "zone": "sysa.xyz."
          },
          "private": "bnVsbA=="
        }
      ]
    }
  ]
}
$
```

### 1. Changing the value of `records` to `[]`:

```
$ cat terraform.tfstate
{
  "version": 4,
  "terraform_version": "0.12.7",
  "serial": 48,
  "lineage": "5d15eb7d-2f44-4172-ac9d-494cb2534cd2",
  "outputs": {},
  "resources": [
    {
      "mode": "managed",
      "type": "powerdns_record",
      "name": "foobar",
      "provider": "provider.powerdns",
      "instances": [
        {
          "schema_version": 0,
          "attributes": {
            "id": "something.sysa.xyz.:::A",
            "name": "something.sysa.xyz.",
            "records": [
              "127.0.0.1"
            ],
            "set_ptr": null,
            "ttl": 300,
            "type": "A",
            "zone": "sysa.xyz."
          },
          "private": "bnVsbA=="
        }
      ]
    }
  ]
}
$ curl -X GET -H "X-Api-Key: secret" http://localhost:8081/api/v1/servers/localhost/zones/sysa.xyz. -s  | jq '.rrsets[].name | match("something.sysa.xyz.")'
{
  "offset": 0,
  "length": 19,
  "string": "something.sysa.xyz.",
  "captures": []
}
$ cat test.tf
resource "powerdns_record" "foobar" {
  zone    = "sysa.xyz."
  name    = "something.sysa.xyz."
  type    = "A"
  ttl     = 300
  records = []
}
$ terraform apply  -auto-approve
powerdns_record.foobar: Refreshing state... [id=something.sysa.xyz.:::A]
powerdns_record.foobar: Destroying... [id=something.sysa.xyz.:::A]
powerdns_record.foobar: Destruction complete after 0s
powerdns_record.foobar: Creating...
powerdns_record.foobar: Creation complete after 0s [id=something.sysa.xyz.:::A]

Apply complete! Resources: 1 added, 0 changed, 1 destroyed.
$ cat terraform.tfstate
{
  "version": 4,
  "terraform_version": "0.12.7",
  "serial": 51,
  "lineage": "5d15eb7d-2f44-4172-ac9d-494cb2534cd2",
  "outputs": {},
  "resources": [
    {
      "mode": "managed",
      "type": "powerdns_record",
      "name": "foobar",
      "provider": "provider.powerdns",
      "instances": [
        {
          "schema_version": 0,
          "attributes": {
            "id": "something.sysa.xyz.:::A",
            "name": "something.sysa.xyz.",
            "records": null,
            "set_ptr": null,
            "ttl": 300,
            "type": "A",
            "zone": "sysa.xyz."
          },
          "private": "bnVsbA=="
        }
      ]
    }
  ]
}
$ curl -X GET -H "X-Api-Key: secret" http://localhost:8081/api/v1/servers/localhost/zones/sysa.xyz. -s  | jq '.rrsets[].name | match("something.sysa.xyz.")'
```

### 2. Changing the value of `records` to `[""]` (same conditions/environment as 1):

```
$ cat terraform.tfstate
{
  "version": 4,
  "terraform_version": "0.12.7",
  "serial": 54,
  "lineage": "5d15eb7d-2f44-4172-ac9d-494cb2534cd2",
  "outputs": {},
  "resources": [
    {
      "mode": "managed",
      "type": "powerdns_record",
      "name": "foobar",
      "provider": "provider.powerdns",
      "instances": [
        {
          "schema_version": 0,
          "attributes": {
            "id": "something.sysa.xyz.:::A",
            "name": "something.sysa.xyz.",
            "records": [
              "127.0.0.1"
            ],
            "set_ptr": null,
            "ttl": 300,
            "type": "A",
            "zone": "sysa.xyz."
          },
          "private": "bnVsbA=="
        }
      ]
    }
  ]
}
$ curl -X GET -H "X-Api-Key: secret" http://localhost:8081/api/v1/servers/localhost/zones/sysa.xyz. -s  | jq '.rrsets[].name | match("something.sysa.xyz.")'
{
  "offset": 0,
  "length": 19,
  "string": "something.sysa.xyz.",
  "captures": []
}
$ cat test.tf
resource "powerdns_record" "foobar" {
  zone    = "sysa.xyz."
  name    = "something.sysa.xyz."
  type    = "A"
  ttl     = 300
  records = [""]
}
$ terraform apply  -auto-approve
powerdns_record.foobar: Refreshing state... [id=something.sysa.xyz.:::A]
powerdns_record.foobar: Destroying... [id=something.sysa.xyz.:::A]
powerdns_record.foobar: Destruction complete after 0s
powerdns_record.foobar: Creating...

Error: Failed to create PowerDNS Record: Error creating record set: something.sysa.xyz.:::A, reason: "Record something.sysa.xyz./A '': Parsing record content (try 'pdnsutil check-zone'): missing field at the end of record content ''"

  on test.tf line 1, in resource "powerdns_record" "foobar":
   1: resource "powerdns_record" "foobar" {


$ cat terraform.tfstate
{
  "version": 4,
  "terraform_version": "0.12.7",
  "serial": 56,
  "lineage": "5d15eb7d-2f44-4172-ac9d-494cb2534cd2",
  "outputs": {},
  "resources": []
}
$ curl -X GET -H "X-Api-Key: secret" http://localhost:8081/api/v1/servers/localhost/zones/sysa.xyz. -s  | jq '.rrsets[].name | match("something.sysa.xyz.")'
$
```

If the suggestion from [here](https://github.com/terraform-providers/terraform-provider-powerdns/issues/31#issuecomment-528086027) is applied, the behavior is:

A. Using `records = []`

```
$ terraform apply  -auto-approve
powerdns_record.foobar: Refreshing state... [id=something.sysa.xyz.:::A]
powerdns_record.foobar: Destroying... [id=something.sysa.xyz.:::A]
powerdns_record.foobar: Destruction complete after 0s
powerdns_record.foobar: Creating...

Error: Couldn't fetch PowerDNS Record: Unknown record ID format

  on test.tf line 1, in resource "powerdns_record" "foobar":
   1: resource "powerdns_record" "foobar" {


$ cat terraform.tfstate
{
  "version": 4,
  "terraform_version": "0.12.7",
  "serial": 42,
  "lineage": "5d15eb7d-2f44-4172-ac9d-494cb2534cd2",
  "outputs": {},
  "resources": []
}
$ curl -X GET -H "X-Api-Key: secret" http://localhost:8081/api/v1/servers/localhost/zones/sysa.xyz. -s  | jq '.rrsets[].name | match("something.sysa.xyz.")'
$
```

B. Using `records = [""]`:

```
$ cat terraform.tfstate
{
  "version": 4,
  "terraform_version": "0.12.7",
  "serial": 44,
  "lineage": "5d15eb7d-2f44-4172-ac9d-494cb2534cd2",
  "outputs": {},
  "resources": [
    {
      "mode": "managed",
      "type": "powerdns_record",
      "name": "foobar",
      "provider": "provider.powerdns",
      "instances": [
        {
          "schema_version": 0,
          "attributes": {
            "id": "something.sysa.xyz.:::A",
            "name": "something.sysa.xyz.",
            "records": [
              "127.0.0.1"
            ],
            "set_ptr": null,
            "ttl": 300,
            "type": "A",
            "zone": "sysa.xyz."
          },
          "private": "bnVsbA=="
        }
      ]
    }
  ]
}
$ curl -X GET -H "X-Api-Key: secret" http://localhost:8081/api/v1/servers/localhost/zones/sysa.xyz. -s  | jq '.rrsets[].name | match("something.sysa.xyz.")'
{
  "offset": 0,
  "length": 19,
  "string": "something.sysa.xyz.",
  "captures": []
}
$ cat test.tf
resource "powerdns_record" "foobar" {
  zone    = "sysa.xyz."
  name    = "something.sysa.xyz."
  type    = "A"
  ttl     = 300
  records = [""]
}
$ terraform apply  -auto-approve
powerdns_record.foobar: Refreshing state... [id=something.sysa.xyz.:::A]
powerdns_record.foobar: Destroying... [id=something.sysa.xyz.:::A]
powerdns_record.foobar: Destruction complete after 0s
powerdns_record.foobar: Creating...

Error: Failed to create PowerDNS Record: Error creating record set: something.sysa.xyz.:::A, reason: "Record something.sysa.xyz./A '': Parsing record content (try 'pdnsutil check-zone'): missing field at the end of record content ''"

  on test.tf line 1, in resource "powerdns_record" "foobar":
   1: resource "powerdns_record" "foobar" {


$ cat terraform.tfstate
{
  "version": 4,
  "terraform_version": "0.12.7",
  "serial": 46,
  "lineage": "5d15eb7d-2f44-4172-ac9d-494cb2534cd2",
  "outputs": {},
  "resources": []
}
$ curl -X GET -H "X-Api-Key: secret" http://localhost:8081/api/v1/servers/localhost/zones/sysa.xyz. -s  | jq '.rrsets[].name | match("something.sysa.xyz.")'
$
```

**Comment by splashx** [comment]
Upon analysis of the issue, among other things we have found that yes, PowerDNS API allows for two ways for a record to be deleted and [that is documented](https://doc.powerdns.com/authoritative/http-api/zone.html#record). 

* Option A: using `changetype: DELETE`

```
{
  "rrsets": [
    {
      "name": "something.sysa.xyz.",
      "type": "A",
      "changetype": "DELETE",
      "records": [""]
    }
  ]
}
```

* Option B: using `changetype: REPLACE` and `records: []`

```
{
  "rrsets": [
    {
      "name": "something.sysa.xyz.",
      "type": "A",
      "changetype": "REPLACE",
      "records": [],
      "comments": [""],
      "ttl": 360
    },
  ]
}
```

However, if we're considering the terraform interface and _declarative infrastructure_, creating and deleting a record is simply a matter of:

* Create 

```hcl
resource "powerdns_record" "foobar" {
  zone    = "sysa.xyz."
  name    = "something.sysa.xyz."
  type    = "A"
  ttl     = 300
  records = ["127.0.0.1"]
}
```

* Delete

```hcl
# resource "powerdns_record" "foobar" {
#   zone    = "sysa.xyz."
#   name    = "something.sysa.xyz."
#   type    = "A"
#   ttl     = 300
#   records = ["127.0.0.1"]
# }
```

Hence, there is actually no need to support `[]` or `[""]` in `records` if the sole purpose is to delete a record. 

But to understand the behavior described in the issue, we need to first understand that we don't really have a 1:1 mapping of `records` (in `resource` definition) to `records` in the API definition. The mapping is actually `records` (in `resource` definition) to `content` (in the API definition):

![Screenshot 2019-09-05 at 13 52 24](https://user-images.githubusercontent.com/1154133/64339963-60342000-cfe5-11e9-8bc2-ad0af25916bc.png)

The code in question is here: 
https://github.com/terraform-providers/terraform-provider-powerdns/blob/3d8c32c5ffe7c6e7932886a86f94d7e51f94eef9/powerdns/resource_powerdns_record.go#L79-L81

And, if one would want to create a record with empty content:

```json
{
  "rrsets": [
    {
      "name": "something.sysa.xyz.",
      "type": "A",
      "changetype": "REPLACE",
      "ttl": 360
      "records": [
        {
          "content": "",
          "disabled": false
        }
      ],

    },
  ]
}
```

PowerDNS will throw an error:

```
{"error": "Record something.sysa.xyz./A '': Parsing record content (try 'pdnsutil check-zone'): missing field at the end of record content ''"}
```

Since we don't _really_ have the possibility to have *actual* empty `records` in the API call (because of the mapping), I vote for _not allowing_ that definition - and throw an error during the pre-check flight. 

And in the same PR I would simply get rid of this block (accordingly):

https://github.com/terraform-providers/terraform-provider-powerdns/blob/3d8c32c5ffe7c6e7932886a86f94d7e51f94eef9/powerdns/resource_powerdns_record.go#L94-L101

**Comment by splashx** [comment]
Some update on this:

`[]` - will not allowed (regardless of record type)
`[""]` - there will be no check for empty record - we leave it to PowerDNS to throw an error. That's because some record Types  _can_ be empty, for example, `TXT` can have content `""`. On the other hand, `A` records, can't and we would see the following error:

```
Error: Failed to create PowerDNS Record: Error creating record set: something3.sysa.xyz.:::A, reason: "Record something3.sysa.xyz./A '': Parsing record content (try 'pdnsutil check-zone'): missing field at the end of record content ''"

```
